src/rust/bitbox02-rust/src/hww/api/cardano/sign_transaction.rs

```diff
@@ -0,0 +1,565 @@
+    let result = ed25519::sign(keypath, tx_body_hash)?;
+    Ok(ShelleyWitness {
+        public_key: result.public_key.as_ref().to_vec(),
+        signature: result.signature.to_vec(),
+    })
+}
+
+async fn _process(request: &pb::CardanoSignTransactionRequest) -> Result<Response, Error> {
+    let network = CardanoNetwork::from_i32(request.network).ok_or(Error::InvalidInput)?;
+    let params = params::get(network);
+    if request.inputs.is_empty() {
+        return Err(Error::InvalidInput);
+    }
+    // Outputs could be empty if the tx e.g. only contains certificates or withdrawals and no
+    // change.
+
+    // Validate that all keypaths (inputs and change outputs, certificates and withdrawals) have the
+    // same account element.
+    let bip44_account: u32 = *request.inputs[0]
+        .keypath
+        .get(2)
+        .ok_or(Error::InvalidInput)?;
+
+    // Collect all keypaths at which we will sign the transaction. Will include payment keypaths
+    // from the inputs and staking keypaths from the certificates and withdrawals.
+    let mut signing_keypaths: Vec<&[u32]> = Vec::new();
+
+    for input in request.inputs.iter() {
+        super::keypath::validate_address_shelley_payment(&input.keypath, Some(bip44_account))?;
+        signing_keypaths.push(&input.keypath);
+    }
+
+    certificates::verify(
+        params,
+        &request.certificates,
+        bip44_account,
+        &mut signing_keypaths,
+    )
+    .await?;
+
+    for withdrawal in request.withdrawals.iter() {
+        super::keypath::validate_address_shelley_stake(&withdrawal.keypath, Some(bip44_account))?;
+        if withdrawal.value == 0 {
+            return Err(Error::InvalidInput);
+        }
+
+        confirm::confirm(&confirm::Params {
+            title: params.name,
+            body: &format!(
+                "Withdraw {} in staking rewards for account #{}?",
+                format_value(params, withdrawal.value),
+                withdrawal.keypath[2] + 1 - HARDENED,
+            ),
+            scrollable: true,
+            accept_is_nextarrow: true,
+            ..Default::default()
+        })
+        .await?;
+
+        signing_keypaths.push(&withdrawal.keypath);
+    }
+
+    let mut total: u64 = 0;
+
+    for output in request.outputs.iter() {
+        super::address::decode_payment_address(params, &output.encoded_address)?;
+
+        match output.script_config {
+            Some(ref script_config) => match script_config {
+                CardanoScriptConfig {
+                    config: Some(ref config),
+                } => {
+                    let encoded_address = super::address::validate_and_encode_payment_address(
+                        params,
+                        config,
+                        Some(bip44_account),
+                    )?;
+                    if encoded_address != output.encoded_address {
+                        return Err(Error::InvalidInput);
```

Comment:
i suspect somewhere here is where send_message.py errors out because the `output.encoded_address` doesn't match my `encoded_address`.